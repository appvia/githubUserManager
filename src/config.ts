export const config = {
  get addUsers(): boolean {
    return process.env.ADD_USERS?.toLowerCase() === 'true'
  },
  get removeUsers(): boolean {
    return process.env.REMOVE_USERS?.toLowerCase() === 'true'
  },
  get exitCodeOnMissmatch(): number {
    return parseInt(process.env.EXIT_CODE_ON_MISMATCH ?? '0') || 0
  },
  get ignoredUsers(): string[] {
    return process.env.IGNORED_USERS?.toLowerCase().split(',') ?? []
  },
  get githubPrivateKey(): string {
    return Buffer.from(process.env.GITHUB_PRIVATE_KEY, 'base64').toString('utf-8')
  },
  get githubAppID(): number {
    return parseInt(process.env.GITHUB_APP_ID)
  },
  get githubInstallationID(): number {
    return parseInt(process.env.GITHUB_INSTALLATION_ID)
  },
  get githubOrg(): string {
    return process.env.GITHUB_ORG ?? ''
  },
  get googleCredentials(): googleCredentials {
    return JSON.parse(Buffer.from(process.env.GOOGLE_CREDENTIALS, 'base64').toString('utf-8'))
  },
  get googleEmailAddress(): string {
    return process.env.GOOGLE_EMAIL_ADDRESS ?? ''
  },
}

export interface googleCredentials {
  type: string
  project_id: string
  private_key_id: string
  private_key: string
  client_email: string
  client_id: string
  auth_uri: string
  token_uri: string
  auth_provider_x509_cert_url: string
  client_x509_cert_url: string
}
