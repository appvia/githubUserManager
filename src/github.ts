import { createAppAuth } from '@octokit/auth-app'
import { Octokit } from '@octokit/rest'
import * as mod from './github'
import { config } from './config'
import { GetResponseTypeFromEndpointMethod } from '@octokit/types'

export function getAuthenticatedOctokit(): Octokit {
  return new Octokit({
    authStrategy: createAppAuth,
    auth: {
      appId: config.githubAppID,
      privateKey: config.githubPrivateKey,
      installationId: config.githubInstallationID,
    },
  })
}

export async function getGithubUsersFromGithub(): Promise<Set<string>> {
  const octokit = mod.getAuthenticatedOctokit()
  const members = await octokit.paginate(octokit.orgs.listMembers, {
    org: config.githubOrg,
  })

  const pendingInvites = await octokit.paginate(octokit.orgs.listPendingInvitations, {
    org: config.githubOrg,
  })
  const pendingGithubAccounts = formatUserList(pendingInvites)

  const githubAccounts = formatUserList(members)

  if (pendingGithubAccounts.size > 0)
    console.log(`Outstanding GitHub invites for ${Array.from(pendingGithubAccounts).join(', ')}`)

  return new Set([...githubAccounts, ...pendingGithubAccounts])
}

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export function formatUserList(users): Set<string> {
  return new Set(
    users
      .map((user) => user.login?.toLowerCase())
      .flat()
      .filter(Boolean),
  )
}

export async function getUserIdFromUsername(username: string): Promise<number> {
  const octokit = mod.getAuthenticatedOctokit()
  console.log(`Looking up user ${username}`)
  let user
  try {
    user = await octokit.users.getByUsername({ username })
  } catch (error) {
    throw `Unable to find user id for ${username}`
  }
  console.log(`User ${username} userid: ${user.data.id}`)
  return user.data.id
}

export async function addUsersToGitHubOrg(users: Set<string>): Promise<void> {
  for (const user of users) {
    await mod.addUserToGitHubOrg(user)
  }
}

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export async function addUserToGitHubOrg(
  user: string,
): Promise<GetResponseTypeFromEndpointMethod<typeof octokit.orgs.createInvitation> | boolean> {
  const octokit = mod.getAuthenticatedOctokit()
  if (config.ignoredUsers.includes(user.toLowerCase())) {
    console.log(`Ignoring add for ${user}`)
    return false
  }
  const userId = await mod.getUserIdFromUsername(user)
  console.log(`Inviting ${user} (${userId} to ${config.githubOrg})`)
  return await octokit.orgs.createInvitation({
    org: config.githubOrg,
    invitee_id: userId,
  })
}

export async function removeUsersFromGitHubOrg(users: Set<string>): Promise<void> {
  for (const user of users) {
    await mod.removeUserFromGitHubOrg(user)
  }
}

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export async function removeUserFromGitHubOrg(
  user: string,
): Promise<GetResponseTypeFromEndpointMethod<typeof octokit.orgs.removeMembershipForUser> | boolean> {
  const octokit = mod.getAuthenticatedOctokit()
  if (config.ignoredUsers.includes(user.toLowerCase())) {
    console.log(`Ignoring remove for ${user}`)
    return false
  }
  console.log(`Removing user/invitation ${user} from ${config.githubOrg}`)
  return octokit.orgs.removeMembershipForUser({
    org: config.githubOrg,
    username: user,
  })
}
