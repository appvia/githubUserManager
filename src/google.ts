import { google } from 'googleapis'
import * as mod from './google'
import { config } from './config'

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export async function googleAuth() {
  const privatekey = config.googleCredentials
  const jwtClient = new google.auth.JWT(
    privatekey.client_email,
    null,
    privatekey.private_key,
    ['https://www.googleapis.com/auth/admin.directory.user.readonly'],
    config.googleEmailAddress,
  )
  await jwtClient.authorize()
  return jwtClient
}

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export async function getAdminService() {
  return google.admin({
    version: 'directory_v1',
    auth: await googleAuth(),
  })
}

export async function getGithubUsersFromGoogle(): Promise<Set<string>> {
  const service = await mod.getAdminService()

  const userList = await service.users.list({
    customer: 'my_customer',
    maxResults: 250,
    projection: 'custom',
    fields: 'users(customSchemas/Accounts/github(value))',
    customFieldMask: 'Accounts',
  })

  const githubAccounts = mod.formatUserList(userList.data.users)
  return githubAccounts
}

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export function formatUserList(users): Set<string> {
  return new Set(
    users
      .map((user) => user.customSchemas?.Accounts?.github?.map((account) => account.value?.toLowerCase()))
      .flat()
      .filter(Boolean),
  )
}
