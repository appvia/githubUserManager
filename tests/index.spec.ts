jest.mock('../src/google')
jest.mock('../src/github')
import * as google from '../src/google'
import * as github from '../src/github'
import * as mod from '../index'

let processExitSpy
let consoleSpy

beforeEach(() => {
  processExitSpy = jest.spyOn(global.process, 'exit').mockImplementation(() => {
    return undefined as never
  })
  consoleSpy = jest.spyOn(global.console, 'log').mockImplementation()
})

describe('missmatch', () => {
  beforeEach(() => {
    // @ts-expect-error mockResolved unexpected
    google.getGithubUsersFromGoogle.mockResolvedValue(new Set(['a', 'd']))
    // @ts-expect-error mockResolved unexpected
    github.getGithubUsersFromGithub.mockResolvedValue(new Set(['b', 'c', 'a']))
  })
  it('should have consistent console output', async () => {
    await mod.run()
    return expect(consoleSpy).toMatchSnapshot()
  })

  it('should exit with 0 by default as there is a missmatch', async () => {
    await mod.run()
    return expect(processExitSpy).toBeCalledWith(0)
  })
  it('should exit with 122 if defined when there is a missmatch', async () => {
    process.env.EXIT_CODE_ON_MISMATCH = '122'
    await mod.run()
    return expect(processExitSpy).toBeCalledWith(122)
  })
  it('should not add users if not set to', async () => {
    delete process.env.ADD_USERS
    await mod.run()
    return expect(github.addUsersToGitHubOrg).not.toBeCalled()
  })
  it('should not remove users if not set to', async () => {
    delete process.env.REMOVE_USERS
    await mod.run()
    return expect(github.removeUsersFromGitHubOrg).not.toBeCalled()
  })
  it('should add users if set to', async () => {
    process.env.ADD_USERS = 'true'
    await mod.run()
    return expect(github.addUsersToGitHubOrg).toMatchSnapshot()
  })
  it('should remove users if set to', async () => {
    process.env.REMOVE_USERS = 'true'
    await mod.run()
    return expect(github.removeUsersFromGitHubOrg).toMatchSnapshot()
  })
  it('should have consistent console output with full destructive mode on', async () => {
    process.env.REMOVE_USERS = 'true'
    process.env.ADD_USERS = 'true'
    await mod.run()
    return expect(consoleSpy).toMatchSnapshot()
  })
})

describe('match', () => {
  beforeEach(() => {
    // @ts-expect-error mockResolved unexpected
    google.getGithubUsersFromGoogle.mockResolvedValue(new Set(['a', 'b', 'c', 'd']))
    // @ts-expect-error mockResolved unexpected
    github.getGithubUsersFromGithub.mockResolvedValue(new Set(['a', 'b', 'c', 'd']))
  })
  it('should have consistent console output', async () => {
    await mod.run()
    return expect(consoleSpy).toMatchSnapshot()
  })
  it('should exit with 0 by default', async () => {
    await mod.run()
    return expect(processExitSpy).toBeCalledWith(0)
  })
  it('should not exit with 122 if defined', async () => {
    process.env.EXIT_CODE_ON_MISMATCH = '122'
    await mod.run()
    return expect(processExitSpy).not.toBeCalledWith(122)
  })
  it('should not add users', async () => {
    process.env.ADD_USERS = 'true'
    await mod.run()
    return expect(github.addUsersToGitHubOrg).not.toBeCalled()
  })
  it('should not remove users', async () => {
    process.env.REMOVE_USERS = 'true'
    await mod.run()
    return expect(github.removeUsersFromGitHubOrg).not.toBeCalled()
  })
})
