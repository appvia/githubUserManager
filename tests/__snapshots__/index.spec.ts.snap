// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`match should have consistent console output 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "Users from google: a, b, c, d",
    ],
    Array [
      "Users from github: a, b, c, d",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`missmatch should add users if set to 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Set {
        "d",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`missmatch should have consistent console output 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "Users from google: a, d",
    ],
    Array [
      "Users from github: b, c, a",
    ],
    Array [
      "Users not in google: b, c",
    ],
    Array [
      "Users not in github: d",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`missmatch should have consistent console output with full destructive mode on 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "Users from google: a, d",
    ],
    Array [
      "Users from github: b, c, a",
    ],
    Array [
      "Users not in google: b, c",
    ],
    Array [
      "Users not in github: d",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`missmatch should remove users if set to 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Set {
        "b",
        "c",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
